sklearn/linear_model/cd_fast.pyx

```diff
@@ -369,169 +600,190 @@ def sparse_enet_coordinate_descent(floating [:] w,
+                        R_shift -= X_mean_ii * d_w_ii
 
                 # update the maximum absolute coefficient update
-                d_w_ii = fabs(w[ii] - w_ii)
+                d_w_ii = fabs(d_w_ii)
                 if d_w_ii > d_w_max:
                     d_w_max = d_w_ii
 
-                if w[ii] > w_max:
-                    w_max = w[ii]
-            if w_max == 0.0 or d_w_max / w_max < d_w_tol or n_iter == max_iter - 1:
-                # the biggest coordinate update of this iteration was smaller than
-                # the tolerance: check the duality gap as ultimate stopping
-                # criterion
-
-                # sparse X.T / dense R dot product
-                if center:
-                    R_sum = 0.0
-                    for jj in range(n_samples):
-                        R_sum += R[jj]
-
-                for ii in range(n_features):
-                    X_T_R[ii] = 0.0
-                    for jj in range(X_indptr[ii], X_indptr[ii + 1]):
-                        X_T_R[ii] += X_data[jj] * R[X_indices[jj]]
-
-                    if center:
-                        X_T_R[ii] -= X_mean[ii] * R_sum
-                    XtA[ii] = X_T_R[ii] - beta * w[ii]
+                if fabs(w[ii]) > w_max:
+                    w_max = fabs(w[ii])
 
-                if positive:
-                    dual_norm_XtA = max(n_features, &XtA[0])
-                else:
-                    dual_norm_XtA = abs_max(n_features, &XtA[0])
+            # Variable screening
+            if (w_max == 0.0 or
+                d_w_max / w_max < d_w_tol or
+                n_iter == max_iter - 1 or
+                screening > 0 and n_iter % screening == 0) and n_iter != 0:
+                # the biggest coordinate update of this iteration was smaller
+                # than the tolerance: check the duality gap as ultimate
+                # stopping criterion
+                gap = sparse_enet_duality_gap(n_samples, n_features,
+                                              <floating*> X_data.data,
+                                              <int*> X_indices.data,
+                                              <int*> X_indptr.data, X_mean,
+                                              <floating*> y.data, y_sum,
+                                              R, R_shift, w, XtA, X_T_R,
+                                              dual_scaling, alpha, beta,
+                                              positive, center, disabled,
+                                              n_features - n_active)
+
+                # Break if we reached desired tolerance
+                if gap < tol:
+                    n_iter -= 1
+                    break
 
-                # R_norm2 = np.dot(R, R)
-                R_norm2 = dot(n_samples, &R[0], 1, &R[0], 1)
+                # Remove inactive variables (screening)
+                if screening > 0:
+                    n_active = 0
+                    r_screening = sqrt(2. * gap) / alpha
```

Comment:
So we use spheres instead of domes, I reckon that both perform similarly and sphere are easier to implement ?